### DISCUSSION TITLE
classes/class_audiostreaminteractive

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#79] [User-contributed notes]
# classes/class_audiostreaminteractive

Inherits: AudioStream< Resource< RefCounted< Object Audio stream that can playback music interactively, combining clips and a transition table. Description: This is an audio stream that can playbac...

https://docs.godotengine.org/en/latest/classes/class_audiostreaminteractive.html

**Comment by andersonbelli** [comment]
For anyone wondering how to switch between clips you can simple use the AudioStream _switch_to_clip_ parameter and the name of the clip you want to switch to, example:

`main_audio["parameters/switch_to_clip"] = "shop_music"`

**Reply by tot121212** [reply]
This appears to be the only way to switch clips.

**Reply by tot121212** [reply]
You can also use the syntax
`main_audio.set("parameters/switch_to_clip", "shop_music")`
or
`main_audio.get("parameters/example_parameter")`
which I find to be cleaner

**Comment by GamplaxPramgrammer** [comment]
Another way to switch between clips is to use get_stream_playback().switch_to_clip() or get_stream_playback().switch_to_clip_by_name(); which are functions of the AudioStreamPlayer class and not the AudioStreamInteractive resource.

**Reply by tot121212** [reply]
Can't find either of these functions on the AudioStreamPlayer, nor the AudioStreamInteractive

**Reply by GamplaxPramgrammer** [reply]
That's because it doesn't autofill for you to use the function... for some reason.

**Reply by jwel** [reply]
I noticed that AudioStreamPlaybackInteractive derives from AudioStreamPlayback and is what you get back from calling get_stream_playback() on the AudioStreamInteractive. The AudioStreamPlaybackInteractive class has switch_to_clip and switch_to_clip_interactive methods: https://docs.godotengine.org/en/stable/classes/class_audiostreamplaybackinteractive.html

**Reply by jwel** [reply]
Explicitly setting type to AudioStreamPlaybackInteractive will help the editor auto-complete its methods. e.g.
```gdscript
@onready var music_interactive_player := $MusicBank/AudioStreamPlayer_MusicInteractive

func transition_music_to_campfire_scene() -> void:
	var playback: AudioStreamPlaybackInteractive = music_interactive_player.get_stream_playback()
	playback.switch_to_clip_by_name("MyClipName")
```

**Comment by p3drac0** [comment]
Hi everyone!!

I was trying different things with the AudioStreamInteractive and the transitions but I haven't been able to do this yet:

I would like to make a transition with fade out in which second clip doesn't start until first one is over. The problem is that if you put the fade in mode, the second clip starts at the same time that the first one starts the fade out, in a abrupt way. I've tried with every params but i haven't got to the solution yet.

Do anyone know to solve this?

**Comment by voylin** [comment]
I can't figure this one out at all, how do we add clips to this from code? Or is this not supported for some reason?

Edit: Found out. First you increase the clip_count manually, then you use the set_clip_stream and set_clip_name

**Comment by Keyvan-Rostyslav** [comment]
I wrote in C#, don't have to thank me =)

```csharp
/// <summary>
/// The method is used to change the audio clip by name
/// </summary>
/// <param name="audioSource">Audio Source that performs audio playback</param>
/// <param name="clipName">Name of next clip</param>

private void TransitionInteractionAudioClip(AudioStreamPlayer audioSource, string clipName)
{
    var playBack = audioSource.GetStreamPlayback();

    if(playBack != null && playBack is AudioStreamPlaybackInteractive playBackInteractive)
    {
        playBackInteractive.SwitchToClipByName(clipName);
    }
    else
    {
        GD.PrintErr($"{Name} : The audio source does not have an audio clip or an invalid data type is specified that differs from AudioStreamInteractive");
    }
}

**Comment by Dr-Bamboo-Dev** [comment]
Are all the audio clips loaded the whole time, even when they're not playing? If so, there is still a use for a manually coded crossfade between levels that unloads old groups of tracks.
